### PR TITLE
Helpful Startup Errors

### LINKED ISSUE #1349: Waiting for the website to be up before starting in development mode
# Abstract

When the development mode is enabled, the bot should wait for the website to be up before continuing the startup process.

# Rationale

With the current setup using Docker Compose, the bot will start making requests to the website before the website even starts accepting requests, which will cause the bot to fail to start. 

# Specification

We can have a loop trying to ping the server URL every second for a short duration before moving on. Note that the bot should fail to start after some time.

### PR DESCRIPTION
**Pull request by HassanAbouelela**
# Description
One of the biggest hurdles that people face when setting up the bot, is the lack of helpful information when things go wrong. From redis, to the API, plenty can go wrong, and it requires experience to read the traceback and figure out why.

Closes #1349.

# Inspiration
This PR aims to take a step in fixing that, by offering more descriptive errors on startup. This is inspired by #1385, but only covers a portion of the proposed changes.

# Implementation
This PR does two things:
1. It logs a message for redis and API errors, so we don't keep getting confused people asking why they got a 100 line traceback when they hit run.
2. It adds retries to site connection, so it solves the problem of the bot trying to connect to the site before it warms up in docker. (Right now that is at 3 attempts with 5 seconds between each).

### CODE REVIEW

File: bot/bot.py [hunk 1]
```diff
@@ -19,6 +19,14 @@
 LOCALHOST = "127.0.0.1"
 
 
+class StartupError(Exception):
+    """Exception class for startup errors."""
+
+    def __init__(self, base: Exception):
+        super()
```
Comment (Akarys42) [inline bot/bot.py:26]: I believe you meant
```suggestion
        super().__init__()
```

File: bot/bot.py [hunk 2]
```diff
@@ -81,6 +89,22 @@ async def cache_filter_list_data(self) -> None:
...
+        while True:
+            try:
+                log.info(f"Attempting site connection: {attempts + 1}/{constants.URLs.connect_max_retries}")
+                await self.api_client.get("healthcheck")
+                break
+
+            except (aiohttp.ClientConnectorError, aiohttp.ServerDisconnectedError) as e:
+                attempts += 1
+                if attempts == constants.URLs.connect_max_retries:
+                    raise e
```
Comment (Akarys42) [inline bot/bot.py:105]: Nitpicky, but it can just be
```suggestion
                    raise
```

File: bot/__main__.py [hunk 3]
```diff
@@ -1,10 +1,26 @@
...
+    message = "Unknown Startup Error Occurred."
+    if type(e.exception) in [aiohttp.ClientConnectorError, aiohttp.ServerDisconnectedError]:
+        message = "Could not connect to site API. Is it running?"
+    elif isinstance(e.exception, OSError):
+        message = "Could not connect to Redis. Is it running?"
+
+    # The exception is logged with an empty message so the actual message is visible at the bottom
+    log = logging.getLogger("bot")
+    log.fatal("", exc_info=e.exception)
+    log.fatal(message)
```
Comment (Akarys42) [inline bot/__main__.py:26]: This should exit with a non-zero code

File: bot/__main__.py [hunk 4]
```diff
@@ -1,10 +1,26 @@
...
-bot.instance = Bot.create()
-bot.instance.load_extensions()
-bot.instance.run(constants.Bot.token)
+try:
+    bot.instance = Bot.create()
+    bot.instance.load_extensions()
+    bot.instance.run(constants.Bot.token)
+except StartupError as e:
+    message = "Unknown Startup Error Occurred."
+    if type(e.exception) in [aiohttp.ClientConnectorError, aiohttp.ServerDisconnectedError]:
```
Comment (Akarys42) [inline bot/__main__.py:18]: This should be an `isinstance` check ~~and it should be a tuple, not a list :P~~

**Review by Akarys42 — CHANGES_REQUESTED** [review]
This seems to be working fine, although I have some comments to make

**Review by Akarys42 — APPROVED** [review]
LGTM, thanks!

**Review by Shivansh-007 — APPROVED** [review]
Thanks! Looks Good To Me!

**Pull request merged**